hls4ml/templates/vivado/nnet_utils/nnet_graph.h

```diff
@@ -0,0 +1,439 @@
+    #pragma HLS ARRAY_PARTITION variable=data2_logits complete dim=0
+    nnet::dense_resource<data_T, data_T, typename CONFIG_T::dense_config3>(data1, data2_logits, weights2, biases2);
+    data_T data2[CONFIG_T::dense_config3::n_out];
+    #pragma HLS ARRAY_PARTITION variable=data2 complete dim=0
+    nnet::relu<data_T, data_T, typename CONFIG_T::relu_config3>(data2_logits, data2);
+
+    nnet::dense_resource<data_T, res_T, typename CONFIG_T::dense_config4>(data2, res, weights3, biases3);
+  }
+
+  template<class data_T, class index_T, class res_T, typename CONFIG_T>
+    void aggregate(
+            data_T    edge_attr_1D[CONFIG_T::n_edge*CONFIG_T::edge_dim],
+            index_T   edge_index_1D[CONFIG_T::n_edge*2],
+            res_T     edge_attr_aggr_1D[CONFIG_T::n_node*CONFIG_T::edge_dim])
+  {
+    //initialize arrays
+    // 1. edge_attr (input)
+    data_T edge_attr[CONFIG_T::n_edge][CONFIG_T::edge_dim];
+    #pragma HLS ARRAY_PARTITION variable=edge_attr complete dim=0
+    nnet::vec_to_mat<data_T, data_T, typename CONFIG_T::edge_attr_config>(edge_attr_1D, edge_attr);
+
+    // 2. edge_index (input)
+    index_T edge_index[CONFIG_T::n_edge][2];
+    #pragma HLS ARRAY_PARTITION variable=edge_index complete dim=0
+    nnet::vec_to_mat<index_T, index_T, typename CONFIG_T::edge_index_config>(edge_index_1D, edge_index);
+
+    //3. num_edge_per_node (intermediate), 4. edge_aggr_mask (intermediate)
+    index_T num_edge_per_node[CONFIG_T::n_node];
+    #pragma HLS ARRAY_PARTITION variable=num_edge_per_node complete dim=0
+    ap_uint<1> edge_aggr_mask[CONFIG_T::n_node];
+    #pragma HLS ARRAY_PARTITION variable=edge_aggr_mask complete dim=0
+    for(int i=0; i<CONFIG_T::n_node; i++){
+      #pragma HLS UNROLL
+      num_edge_per_node[i] = 0;
+      if(CONFIG_T::aggr==aggr_max){
+        edge_aggr_mask[i] = 0;
+      }
+    }
+
+    //4. edge_attr_aggr (output)
+    res_T edge_attr_aggr[CONFIG_T::n_node][CONFIG_T::edge_dim];
+    #pragma HLS ARRAY_PARTITION variable=edge_update_aggr complete dim=0
+    if((CONFIG_T::aggr==aggr_sum)||(CONFIG_T::aggr==aggr_mean)){
+      for(int i=0; i < CONFIG_T::n_node; i++){
+        for(int j=0; j<CONFIG_T::edge_dim; j++){
+          #pragma HLS UNROLL
+          edge_attr_aggr[i][j] = 0;
+        }
+      }
+    }
+    else{ //CONFIG_T:aggr==aggr_max, we want to initialize this with the most negative number we can represent
+      // note this only works for ap_ufixed types
+      res_T most_negative_num;
+      most_negative_num.V = 1;
+      most_negative_num.V <<= res_T::width - 1;
+
+      for(int i=0; i < CONFIG_T::n_node; i++){
+        for(int j=0; j<CONFIG_T::edge_dim; j++){
+          #pragma HLS UNROLL
+          edge_attr_aggr[i][j] = most_negative_num;
+        }
+      }
+    }
+
+    int receiver_col;
+    if(CONFIG_T::flow == source_to_target){
+      receiver_col = 1;
+    }
+    else{
+      receiver_col = 0;
+    }
+
+    #pragma HLS PIPELINE II=CONFIG_T::reuse_factor

```

Comment:
So I tested a few different implementations of this using the basic configuration (`--max-nodes=28 --max-edges=51 --precision=ap_fixed<16,8> --reuse=1`):
1. Removing the function pipeline pragma
2. Removing all the unroll pragmas 
3. Removing the function pipeline pragma and all the unroll pragmas 
4. Pipelining the main edge-loop, removing the unroll pragmas from the main edge-loop, and removing the function pipeline pragma

These all yielded identical syntheses as the baseline (what we currently have)... which one should we go with?

Reply:
What I meant to say was that (I think) having the PIPELINE pragma here will cause the nodes loop to be fully unrolled, regardless of the `reuse_factor`. But you'd typically have a few tens-hundreds of nodes for a graph network problem to be useful, so unrolling the nodes loop will cause an explosion of resource usage.
But then maybe I'm missing some crucial point here. Should we have a quick chat so that I can give more useful comments?